[allure-testng] Could not write Allure test result

from https://github.com/allure-framework/allure2/issues/738

[//]: # (
. Note: for support questions, please use Stackoverflow or Gitter**. 
. This repository's issues are reserved for feature requests and bug reports.
.
. In case of any problems with Allure Jenkins plugin** please use the following repository 
. to create an issue: https://github.com/jenkinsci/allure-plugin/issues
.
. Make sure you have a clear name for your issue. The name should start with a capital 
. letter and no dot is required in the end of the sentence. An example of good issue names:
.
. - The report is broken in IE11
. - Add an ability to disable default plugins
. - Support emoji in test descriptions
)

#### I'm submitting a ... 
  - [x] bug report
  - [ ] feature request
  - [ ] support request => Please do not submit support request here, see note at the top of this template.

#### What is the current behavior?


#### If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem
In my testcase,if run failed,it will throw AllureResultsWriteException

#### What is the expected behavior?


#### What is the motivation / use case for changing the behavior?


#### Please tell us about your environment:

| Allure version | 2.2.0 |
| --- | --- |
| Test framework | testng@6.12 |
| Allure adaptor | allure-testng@2.0-BETA21 |
| Generate report using | allure-maven@2.19 |


#### Other information
 `
 io.qameta.allure.AllureResultsWriteException: Could not write Allure test result
 `
 `
	at io.qameta.allure.FileSystemResultsWriter.write(FileSystemResultsWriter.java:43)
 `
 `
	at io.qameta.allure.AllureLifecycle.lambda$writeTestCase$12(AllureLifecycle.java:197)
 `
 `
	at java.util.Optional.ifPresent(Optional.java:159)
 `
 `
	at io.qameta.allure.AllureLifecycle.writeTestCase(AllureLifecycle.java:195)
 `
 `
	at io.qameta.allure.testng.AllureTestNg.onTestFailure(AllureTestNg.java:260)
 `
 `
	at org.testng.internal.Invoker.runTestListeners(Invoker.java:1738)
  `
 `
	at org.testng.internal.Invoker.runTestListeners(Invoker.java:1722)
  `
 `
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1258)
  `
 `
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
 `
 `
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:109)
 `
 `
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
 `
 `
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
 `
 `
	at java.lang.Thread.run(Thread.java:748)
 `
 `
Caused by: java.nio.channels.ClosedByInterruptException
  `
 `
	at java.nio.channels.spi.AbstractInterruptibleChannel.end(AbstractInterruptibleChannel.java:202)
 `
 `
	at sun.nio.ch.FileChannelImpl.write(FileChannelImpl.java:216)
  `
 `
	at java.nio.channels.Channels.writeFullyImpl(Channels.java:78)
 `
 `
	at java.nio.channels.Channels.writeFully(Channels.java:101)
 `
 `
	at java.nio.channels.Channels.access$000(Channels.java:61)
  `
 `
	at java.nio.channels.Channels$1.write(Channels.java:174)
 `
 `
	at com.fasterxml.jackson.core.json.UTF8JsonGenerator._flushBuffer(UTF8JsonGenerator.java:2085)
  `
 `
	at com.fasterxml.jackson.core.json.UTF8JsonGenerator.close(UTF8JsonGenerator.java:1127)
 `
 `
	at com.fasterxml.jackson.databind.ObjectMapper._configAndWriteValue(ObjectMapper.java:3898)
  `
 `
	at com.fasterxml.jackson.databind.ObjectMapper.writeValue(ObjectMapper.java:3164)
  `
 `
	at io.qameta.allure.FileSystemResultsWriter.write(FileSystemResultsWriter.java:41)
 `
 `
	... 12 more
`


<!-- Love allure-report? Please consider supporting our collective:
👉  https://opencollective.com/allure-report/donate -->

I added to my project: 
`        <dependency>
        <groupId>com.fasterxml.jackson.core</groupId>
        <artifactId>jackson-annotations</artifactId>
        <version>2.9.5</version>
    </dependency>
        <!-- https://mvnrepository.com/artifact/com.fasterxml.jackson.core/jackson-core -->
        <dependency>
            <groupId>com.fasterxml.jackson.core</groupId>
            <artifactId>jackson-core</artifactId>
            <version>2.9.5</version>
        </dependency>
        <!-- https://mvnrepository.com/artifact/com.fasterxml.jackson.core/jackson-databind -->
        <dependency>
            <groupId>com.fasterxml.jackson.core</groupId>
            <artifactId>jackson-databind</artifactId>
            <version>2.9.5</version>
        </dependency>`

However issue is still on going when running tests in parallel after retry:

> `Exception in thread "TestNG-test=Test-6" io.qameta.allure.AllureResultsWriteException: Could not write Allure test result
	at io.qameta.allure.FileSystemResultsWriter.write(FileSystemResultsWriter.java:43)
	at io.qameta.allure.AllureLifecycle.lambda$writeTestCase$12(AllureLifecycle.java:197)
	at java.util.Optional.ifPresent(Optional.java:159)
	at io.qameta.allure.AllureLifecycle.writeTestCase(AllureLifecycle.java:195)
	at io.qameta.allure.testng.AllureTestNg.onTestFailure(AllureTestNg.java:272)
	at org.testng.internal.TestListenerHelper.runTestListeners(TestListenerHelper.java:67)
	at org.testng.internal.Invoker.runTestListeners(Invoker.java:1389)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1042)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:109)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.nio.channels.ClosedByInterruptException
	at java.nio.channels.spi.AbstractInterruptibleChannel.end(AbstractInterruptibleChannel.java:202)
	at sun.nio.ch.FileChannelImpl.write(FileChannelImpl.java:216)
	at java.nio.channels.Channels.writeFullyImpl(Channels.java:78)
	at java.nio.channels.Channels.writeFully(Channels.java:101)
	at java.nio.channels.Channels.access$000(Channels.java:61)
	at java.nio.channels.Channels$1.write(Channels.java:174)
	at com.fasterxml.jackson.core.json.UTF8JsonGenerator._flushBuffer(UTF8JsonGenerator.java:2085)
	at com.fasterxml.jackson.core.json.UTF8JsonGenerator._writeStringSegment2(UTF8JsonGenerator.java:1400)
	at com.fasterxml.jackson.core.json.UTF8JsonGenerator._writeStringSegment(UTF8JsonGenerator.java:1347)
	at com.fasterxml.jackson.core.json.UTF8JsonGenerator._writeStringSegments(UTF8JsonGenerator.java:1228)
	at com.fasterxml.jackson.core.json.UTF8JsonGenerator.writeString(UTF8JsonGenerator.java:453)
	at com.fasterxml.jackson.databind.ser.std.StringSerializer.serialize(StringSerializer.java:41)
	at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:727)
	at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:719)
	at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:155)
	at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:727)
	at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:719)
	at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:155)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:480)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:319)
	at com.fasterxml.jackson.databind.ObjectMapper._configAndWriteValue(ObjectMapper.java:3893)
	at com.fasterxml.jackson.databind.ObjectMapper.writeValue(ObjectMapper.java:3164)
	at io.qameta.allure.FileSystemResultsWriter.write(FileSystemResultsWriter.java:41)
`

Full stack trace:
*
at io.qameta.allure.FileSystemResultsWriter.write(FileSystemResultsWriter.java:43)
10:59:48 	at io.qameta.allure.AllureLifecycle.writeTestCase(AllureLifecycle.java:400)
10:59:48 	at io.qameta.allure.testng.AllureTestNg.onTestFailure(AllureTestNg.java:272)
10:59:48 	at org.testng.internal.TestListenerHelper.runTestListeners(TestListenerHelper.java:67)
10:59:48 	at org.testng.internal.Invoker.runTestListeners(Invoker.java:1389)
10:59:48 	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1042)
10:59:48 	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
10:59:48 	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:109)
10:59:48 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
10:59:48 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
10:59:48 	at java.lang.Thread.run(Thread.java:748)
10:59:48 Caused by: java.nio.channels.ClosedByInterruptException
10:59:48 	at java.nio.channels.spi.AbstractInterruptibleChannel.end(AbstractInterruptibleChannel.java:202)
10:59:48 	at sun.nio.ch.FileChannelImpl.write(FileChannelImpl.java:216)
10:59:48 	at java.nio.channels.Channels.writeFullyImpl(Channels.java:78)
10:59:48 	at java.nio.channels.Channels.writeFully(Channels.java:101)
10:59:48 	at java.nio.channels.Channels.access$000(Channels.java:61)
10:59:48 	at java.nio.channels.Channels$1.write(Channels.java:174)
10:59:48 	at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
10:59:48 	at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
10:59:48 	at java.io.FilterOutputStream.close(FilterOutputStream.java:158)
10:59:48 	at com.fasterxml.jackson.core.json.UTF8JsonGenerator.close(UTF8JsonGenerator.java:1138)
10:59:48 	at com.fasterxml.jackson.databind.ObjectMapper._configAndWriteValue(ObjectMapper.java:3910)
10:59:48 	at com.fasterxml.jackson.databind.ObjectMapper.writeValue(ObjectMapper.java:3176)
10:59:48 	at io.qameta.allure.FileSystemResultsWriter.write(FileSystemResultsWriter.java:41)
10:59:48 	... 10 more
10:59:48 	Suppressed: java.nio.channels.ClosedChannelException
10:59:48 		at sun.nio.ch.FileChannelImpl.ensureOpen(FileChannelImpl.java:110)
10:59:48 		at sun.nio.ch.FileChannelImpl.write(FileChannelImpl.java:199)
10:59:48 		at java.nio.channels.Channels.writeFullyImpl(Channels.java:78)
10:59:48 		at java.nio.channels.Channels.writeFully(Channels.java:101)
10:59:48 		at java.nio.channels.Channels.access$000(Channels.java:61)
10:59:48 		at java.nio.channels.Channels$1.write(Channels.java:174)
10:59:48 		at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
10:59:48 		at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
10:59:48 		at java.io.FilterOutputStream.close(FilterOutputStream.java:158)
10:59:48 		at io.qameta.allure.FileSystemResultsWriter.write(FileSystemResultsWriter.java:42)
10:59:48 		... 10 more*

--- a/allure-java-commons/src/main/java/io/qameta/allure/FileSystemResultsWriter.java
+++ b/allure-java-commons/src/main/java/io/qameta/allure/FileSystemResultsWriter.java
@@ -5,10 +5,8 @@ import io.qameta.allure.model.Allure2ModelJackson;
 import io.qameta.allure.model.TestResult;
 import io.qameta.allure.model.TestResultContainer;
 
-import java.io.BufferedOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
-import java.io.OutputStream;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.Objects;
@@ -35,8 +33,8 @@ public class FileSystemResultsWriter implements AllureResultsWriter {
                 : generateTestResultName(testResult.getUuid());
         createDirectories(outputDirectory);
         final Path file = outputDirectory.resolve(testResultName);
-        try (OutputStream os = new BufferedOutputStream(Files.newOutputStream(file))) {
-            mapper.writeValue(os, testResult);
+        try {
+            mapper.writeValue(file.toFile(), testResult);
         } catch (IOException e) {
             throw new AllureResultsWriteException("Could not write Allure test result", e);
         }
@@ -49,8 +47,8 @@ public class FileSystemResultsWriter implements AllureResultsWriter {
                 : generateTestResultContainerName(testResultContainer.getUuid());
         createDirectories(outputDirectory);
         final Path file = outputDirectory.resolve(testResultContainerName);
-        try (OutputStream os = new BufferedOutputStream(Files.newOutputStream(file))) {
-            mapper.writeValue(os, testResultContainer);
+        try {
+            mapper.writeValue(file.toFile(), testResultContainer);
         } catch (IOException e) {
             throw new AllureResultsWriteException("Could not write Allure test result container", e);
         }

